source/modulo_controllers/include/modulo_controllers/BaseControllerInterface.hpp

```diff
@@ -523,11 +526,32 @@ inline void BaseControllerInterface::set_parameter_value(const std::string& name
 
 template<typename T>
 inline void BaseControllerInterface::add_input(const std::string& name, const std::string& topic_name) {
-  auto buffer = realtime_tools::RealtimeBuffer<std::shared_ptr<modulo_core::EncodedState>>();
-  auto input = ControllerInput(buffer);
-  create_input(input, name, topic_name);
-  input_message_pairs_.insert_or_assign(
-      name, modulo_core::communication::make_shared_message_pair(std::make_shared<T>(), get_node()->get_clock()));
+  if constexpr (modulo_core::concepts::CustomT<T>) {
+    auto buffer = realtime_tools::RealtimeBuffer<T>();
+    auto input = ControllerInput(buffer);
+    auto parsed_name = validate_and_declare_signal(name, "input", topic_name);
+    if (!parsed_name.empty()) {
+      inputs_.insert_or_assign(parsed_name, input);
+      custom_input_configuration_callables_.insert_or_assign(
+          name, [this](const std::string& name, const std::string& topic) {
+            auto subscription =
+                get_node()->create_subscription<T>(topic, qos_, [this, name](const std::shared_ptr<T> message) {
+                  // TODO: should we try catch bad_any_cast in here?
+                  auto buffer =
+                      std::any_cast<realtime_tools::RealtimeBuffer<std::shared_ptr<T>>>(inputs_.at(name).buffer);
+                  buffer.writeFromNonRT(message);
+                  inputs_.at(name).timestamp = std::chrono::steady_clock::now();
+                });
+            subscriptions_.push_back(subscription);
+          });
+    }
+  } else {
+    auto buffer = realtime_tools::RealtimeBuffer<std::shared_ptr<modulo_core::EncodedState>>();
+    auto input = ControllerInput(buffer);
+    create_input(input, name, topic_name);
```

Comment:
There is a problem here (in the existing code). The `name` might be different from the `parsed_name` that we will use to refer to the input so we can't create the message pair with `name` but have to somehow retrieve the parsed name

Reply:
In fact, we should not use `create_input` here but repeat the content from `create_input` and return early if the parsed name is empty